connectors/sources/sharepoint_online.py

```diff
@@ -663,118 +719,259 @@ async def validate_config(self):
-                                list_item_attachment["_id"] = list_item_attachment[
-                                    "odata.id"
-                                ]
-                                list_item_attachment[
-                                    "object_type"
-                                ] = "list_item_attachment"
-                                list_item_attachment["_timestamp"] = list_item[
-                                    "lastModifiedDateTime"
-                                ]
-                                attachment_download_func = partial(
-                                    self.get_attachment_content,
-                                    list_item_attachment,
-                                )
-                                yield list_item_attachment, attachment_download_func
-
-                        download_func = None
-
                         yield list_item, download_func
 
-                async for site_page in self.client.site_pages(site["webUrl"]):
-                    site_page["_id"] = site_page["GUID"]
-                    site_page["object_type"] = "site_page"
+                # Sync site pages
+                async for site_page in self.site_pages(site["webUrl"]):
+                    yield site_page, None
 
-                    for html_field in ["LayoutWebpartsContent", "CanvasContent1"]:
-                        if html_field in site_page:
-                            site_page[html_field] = html_to_text(site_page[html_field])
+    async def get_docs_incrementally(self, sync_cursor, filtering=None):
+        self._sync_cursor = sync_cursor
 
-                    yield site_page, None
+        if self._sync_cursor is None or self._sync_cursor == {}:
+            raise SyncCursorEmpty(
+                "Incremental sync can't start because it doen't have information about the last full sync job. Please run a full sync to solve this."
+            )
+
+        max_data_age = None
+
+        if filtering is not None and filtering.has_advanced_rules():
+            advanced_rules = filtering.get_advanced_rules()
+            max_data_age = advanced_rules["maxDataAge"]
+
+        async for site_collection in self.site_collections():
+            yield site_collection, None, OP_INDEX
+
+            async for site in self.sites(
+                site_collection["siteCollection"]["hostname"],
+                self.configuration["site_collections"],
+            ):
+                yield site, None, OP_INDEX
+
+                async for site_drive in self.site_drives(site["id"]):
+                    yield site_drive, None, OP_INDEX
+
+                    delta_link = self.get_drive_delta_link(site_drive["id"])
+
+                    if not delta_link:
+                        raise DeltaLinkCorrupted(
+                            "Incremental sync can't start because one of the delta tokens is missing or expired. Please run a full sync to solve this."
+                        )
```

Comment:
What about the situation when a new site appeared?

The message says nothing about it and it can be a legit cause.

Also for big customers who do a full sync every several days, it might make sense to make incremental sync pick up newly added site collections, what do you think?

Reply:
Agree, a drive with no cursor should try to retrieve with `root/delta`

Reply:
I agree with you both. What made me do it like that is a possible situation (I don't trust users' input by default) where `sync_cursor` is initiated but doesn't have any delta links in it. In this case, incremental sync job time will be the save as full sync job time.